### DISCUSSION TITLE
Can OAuth login via Json data instead of Form?

### DISCUSSION THREAD

**Discussion by LouisYZK** [fastapi/fastapi#7616] [Questions] [answered]
### Description
I want to use the jwt auth, and learn the tutorial about it. There are sth confused me:
```python
@router.post('/token')
async def login(form_data: OAuth2PasswordRequestForm = Depends()):
    user = await authenticate_user(form_data.username, form_data.password)
    if not user:
        raise HTTPException(status_code=400, detail='incorrect username')
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = await create_access_token(data={'sub': user.username}, expires_delta=access_token_expires)
    return {'access_token': access_token, 'token_type': 'bearer'}
```
If login success, the token will be returned , and next time you send other request need auth like
```python
oauth2_scheme = OAuth2PasswordBearer(tokenUrl='/token')
async def get_current_user(token: str=Depends(oauth2_scheme)):
    pass
```
- How does it works the next request was added a header with Authorization: Bear automically? Where does the token returned stored?
- When I replace the Form with json, the following requests after login are failing to be with Auth Header
```python
async def login(user_data: schemas.UserLogin):
    pass
```
I am sorry about my English and less experience about auth in Web, wish you could solve my confusion. Thanks in advance.

**Comment by dmig-alarstudios** [comment]
Look inside the `OAuth2PasswordRequestForm` -- form fields are described there. You can replace this class with your own, accepting `application/json` instead of `application/x-www-form-urlencoded`.

**Comment by tiangolo** [comment]
Thanks for the help here @dmig-alarstudios ! :clap: :bow: 

Thanks for reporting back and closing the issue @LouisYZK  :+1:

**Comment by bartwroblewski** [comment]
If one makes a custom class as per @dmig-alarstudios  suggestion, is it possible to also (auto-)update the OAuth Swagger form to use `application/json` instead of `www-form-urlencoded`?

**Reply by YuriiMotov** [reply, marked as answer]
I'm afraid, no.
Swagger UI follows the standard. It doesn't use the schema FastAPI app provides.

So, the only way is to duplicate the endpoint: first accepts credentials via Form, second uses JSON (implement the schema taking `OAuth2PasswordRequestForm` as the base and removing `Form`)

**Comment by dmig-alarstudios** [comment]
@bartwroblewski I couldn't find a solution for that ended up with a duplicate endpoint accepting `www-form-urlencoded`

**Comment by yashp280793** [comment]
@dmig-alarstudios I have a requirement where frontend is posting content using content-type = "application/json". Due to OAuth2PasswordRequestForm,I am getting error as "username field required".

Is there a way to add support of application/json or any middleware to convert incoming json to www-form-urlencoded?
I am in situation that either frontend works (with custom class) or swagger works with OAuth2PasswordRequestForm.

**Comment by awiprayitno** [comment]
Just want to add myself also needing application/json because of external requirement. It's a pity that this external vendor does not adhere to oauth2 using form-data, this is the only piece that is different, everything else fits perfectly.

I ended up having a second token route accepting application json and returning {'token': '......'} as required by this external vendor. The rest is the same with oauth2 from the tutorial (really great one by the way ) 
/
